relayer/processor/path_processor.go

```diff
@@ -536,19 +565,63 @@ ChannelHandshakeLoop:
 	}
 }
 
-func (pp *PathProcessor) sendMessages(pathEnd *pathEndRuntime, messages []provider.RelayerMessage) error {
+func (pp *PathProcessor) sendMessages(src *pathEndRuntime, dst *pathEndRuntime, messages []provider.RelayerMessage) error {
 	if len(messages) == 0 {
 		return nil
 	}
-	// TODO construct MsgUpdateClient for this pathEnd, using the latest trusted IBC header from other pathEnd, prepend messages with the MsgUpdateClient, then send the messages to this pathEnd
+	signer, err := dst.chainProvider.Address()
+	if err != nil {
+		return fmt.Errorf("error getting signer address for {%s}: %w", dst.info.ChainID, err)
+	}
+	// If the client state trusted height is not equal to the client trusted state height, we cannot send a MsgUpdateClient
+	// until another block is observed. But we can send the rest without the MsgUpdateClient.
```

Comment:
This is an interesting comment! 

I'm under the impression that every msg submitted to the light client needs to be prepended with a `MsgUpdateClient` unless the msg in question is from further back than the most recent trusted height from whenever the last `MsgUpdateClient` msg was submitted. Perhaps that is what you mean here as well but it's also possible that I'm misunderstanding this bit so would love some clarification!

Reply:
I thought the update client message was not actually needed to relay packets, but we can ensure that the clients never expire by including the update client message with every operation on the client. If we must have the update client to relay packets, then this behavior will need to change.

Reply:
Was digging through the spec and I saw this:
"The light client on the receiving chain, which verifies the sending chain's state, should be updated to the lastest sending chain state if possible. The verification will fail if the latest state of the light client does not include the packet commitment." [[1]](https://github.com/cosmos/ibc-go/blob/main/modules/core/spec/01_concepts.md)

So it sounds like as long as the light clients state is recent enough to include the packet commitment then it's not necessary. So packets that are further back in history don't need an `MsgUpdateClient` but I think since we are working on a block by block basis now it may be necessary to just send one `MsgUpdateClient` per block.

Reply:
So there must be a `MsgUpdateClient` after the `MsgTransfer` in order for the `MsgRecvPacket` on the counterparty to succeed. Seems fine then to enforce that the `MsgUpdateClient` must succeed to avoid hard to debug verification failures.

I had added this to try and increase packet throughput since `MsgUpdateClient` can be sent at most every other block it appears, since they need to wait for the next block's validator set before the next message can be constructed, so will have at least one block gap in between.

But I changed it back to mirror the existing functionality. It can return an error, and then the updated IBC headers in the next block will trigger it again.